File: Atcoder/AtCoder Beginner Contest 168/c++/(Double Dots).cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <string>
#include <set>
#include <queue>
#include <cmath>
#include <algorithm>
#include <bits/stdc++.h>
using namespace std;
#define rep(i,n) for (int i = 0; i < (n); ++i)
#define all(x) (x).begin(),(x).end()
typedef long long ll;
using P = pair<int,int>;
using P2 = pair<ll,P>;
#define INF 1001001001
#define LINF (1LL<<60)
#define F first
#define S second
template<class T> inline bool chmax(T& a, T b){ if(a<b){ a=b; return 1; } return 0; }
template<class T> inline bool chmin(T& a, T b){ if(a>b){ a=b; return 1; } return 0; }

int main() {
  int n, m;
  cin >> n >> m;
  vector<vector<int> > to(n);
  rep(i, m) {
    int a, b;
    cin >> a >> b;
    to[a-1].push_back(b-1);
    to[b-1].push_back(a-1);
  }
  int start = 0;
  queue<int> q;
  vector<int> parent(n, -1);
  parent[0] = 0;
  q.push(start);
  while(!q.empty()) {
    int x = q.front();q.pop();
    for (int nxt : to[x]) {
      if (parent[nxt] != -1) continue;
      parent[nxt] = x;
      q.push(nxt);
    }
  }
  bool flag = true;
  rep(i, n-1) if (parent[i] == -1) flag = false;
  if (flag) {
    cout << "Yes" << endl;
    rep(i, n-1) cout << parent[i+1]+1 << endl;
  } else {
    cout << "No" << endl;
  }
}

File: Atcoder/AtCoder Beginner Contest 168/py/(Double Dots).py
import queue
N, M = map(int, input().split())
nodes = [[] for i in range(N)]
que = queue.Queue()
sign = [-1 for i in range(N)]

for i in range(M):
    a, b = map(int, input().split())
    nodes[a-1].append(b-1)
    nodes[b-1].append(a-1)

sign[0] = 0
que.put(0)

while not que.empty():
    cur = que.get()
    for i in nodes[cur]:
        if sign[i] != -1: continue
        que.put(i)
        sign[i] = cur

if all([i != -1 for i in sign]):
    print("Yes")
    for i in sign[1:]:
        print(i+1)
else:
    print("No")